interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,245 @@
+  private Socket server;
+  private Socket clientSock;
+
+
+  private int bandwidth = DEFAULT_BAND;
+  private int latency = DEFAULT_DELAY_MS;
+
+  public void setBandwidth(int targetBPS) {
+    bandwidth = targetBPS;
+  }
+
+
+  public void setLatency(int targetMillis) {
+    latency = targetMillis / 2; // divide by two because delay is applied in both directions
+  }
+
+  public void setUp() throws IOException {
+    qLength = Math.max(bandwidth * latency / 1000, 1); // convert from ms to s
+    chunkSize = Math.min(qLength, 4 * 1024);
+
+    logger.info("Started new proxy on port " + proxyPort);
+    client = new ServerSocket();
+    client.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = client.accept();
+    server = new Socket();
+    server.connect(new InetSocketAddress(loopback, serverPort));
+
+    DataInputStream clientIn = new DataInputStream(clientSock.getInputStream());
+    DataOutputStream clientOut = new DataOutputStream(server.getOutputStream());
+    DataInputStream serverIn = new DataInputStream(server.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+
+    (new Thread(new ClientReader(clientIn))).start();
+    (new Thread(new ClientWriter(clientOut))).start();
+    (new Thread(new ServerReader(serverIn))).start();
+    (new Thread(new ServerWriter(serverOut))).start();
+
+  }
+
+
+  public void shutDown() throws IOException {
+    logger.info("Proxy shutting down... ");
+    shutDown = true;
+    server.close();
+    client.close();
+  }
+
+  private synchronized void incrementServerQSize(int delta) {
+    serverQSize += delta;
+  }
+
+  private synchronized void incrementClientQSize(int delta) {
+    clientQSize += delta;
+  }
+
+  public static void main(String[] args) throws IOException {
+    TrafficControlProxy p = new TrafficControlProxy();
+    p.setUp();
+  }
+
+  private class ClientReader implements Runnable {
+
+    DataInputStream clientIn;
+
+    public ClientReader(DataInputStream clientIn) {
+      this.clientIn = clientIn;
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = Math.min(chunkSize, qLength - serverQSize);
+          if (nextChunk != 0) {
+            nextChunk = Math.min(chunkSize, qLength - serverQSize);
+            byte[] request = new byte[nextChunk];
+            int readableRequestBytes = clientIn.read(request);
+            long arrivalTime = System.nanoTime() / 1000000;
+            Message m = new Message(arrivalTime, request, readableRequestBytes);
+            clientQueue.add(m);
+            incrementClientQSize(readableRequestBytes);
+          } else {
+            // This is arbitrary right now. Maybe sleep for some fraction of the latency?
+            Thread.sleep(1);
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
```

Comment:
InterruptedException is special.  You need to reset the interrupted bit if you aren't going to handle it.  Printing it is not a good way of handling it.